### DISCUSSION TITLE
New Columns on Relationships doesn't work.

### DISCUSSION THREAD

**Discussion by jurri** [rappasoft/laravel-livewire-tables#1159] [Q&A] [answered]
All new Columns I want make doesn't work.
I cant set a new Name for a Relationship Column...

```
public function columns(): array
    {
        return [
            Column::make('BSID', 'bs_id'),
            Column::make('No', 'detail.no'),
            Column::make('ZIP', 'detail.zip'),
            Column::make('City', 'detail.city'),

// new columns
            Column::make('Firstname', 'detail.firstname'), // <- doesnt work
            Column::make('ZIP', 'detail.firstname') // <- works great 
        ];
    }
```

if I call it without an alias like this `Column::make('detail.firstname')` it crash with "Column not found: 1054 Unknown column 'users.detail.firstname'" --> so it needs an alias on relation

```
select 
  `users`.`id` as `id`, 
  `users`.`bs_id` as `bs_id`, 
  `users_details`.`no` as `detail.no`, 
  `users_details`.`zip` as `detail.zip`, 
  `users_details`.`city` as `detail.city`,
   `users`.`detail`.`firstname` as `detail.firstname`,
  ...
from `users` left join `users_details` on `users_details`.`user_id` = `users`.`id`
```

**Comment by lrljoe** [comment]
Is your column definitely called "firstname" in the database?

Are you eager loading the relationships?

Are you using a builder approach, or the $model approach?

**Reply by jurri** [reply]
Yes, I am sure with the column.
It doesn't matter how the column ist called. They aren't displaying if I change outputnames for table.

Version I use: 
- Laravel: v9.52.5
- Livewire: v2.12.3
- laravel-livewire-tables: v2.12.0

working: 
```
Column::make('No', 'detail.no'),
Column::make('ZIP', 'detail.zip'),
Column::make('City', 'detail.city'),
Column::make('ZIP', 'detail.firstname'), <<<<<<<<
Column::make('Street', 'detail.lastname')  <<<<<<<
```

doesn't work:
```
Column::make('Firstname', 'detail.firstname'),
Column::make('Lastname', 'detail.lastname'),
```

yes relationships loading and working ...
I using building approach. 
```
use App\Models\User;
use Illuminate\Database\Eloquent\Builder;
    public function builder(): Builder
    {
        return User::query();
    }
```
How to use Model with DataTableComponent ? Because ... in my case it is not compatible with the builder() function.

---
other example:
working:
```
Column::make('Street')  // <<<<<<<<<<<<<<<<<<
                ->label(
                    function ($row, Column $column) {
                        return '<button onclick="window.open(\'url/?columnID=' . $row->detail->id . '\')">edit</button>';
                    }
                )->html(),
```
doesn't work:
```
Column::make('Action')  // <<<<<<<<<<<<<<<<<< only name is changed
                ->label(
                    function ($row, Column $column) {
                        return '<button onclick="window.open(\'url/?columnID=' . $row->detail->id . '\')">edit</button>';
                    }
                )->html(),
```

**Reply by lrljoe** [reply]
Builder is definitely the better approach!

If you add 
with('detail') to your builder, do the fields populate?


```php
    public function builder(): Builder
    {
        return User::with('detail')->query();
    }

```

If not, please run the following commands from your shell:
composer update 
php artisan optimize:clear

Let me know if the above make sense!

**Reply by lrljoe** [reply]
If not, if you can share the full component, and models, then I may be able to help further!

**Reply by jurri** [reply, marked as answer]
composer update
php artisan optimize:clear
or `return User::query()->with('detail');`
doesn't help...

but I found how get it displaying :)
I have to write in first Column::make call the title which I want use later.

It has nothing todo with the relationship. But the mistake is only on relationship-columns.
```php
public static function make(**string $title**, string $from = null): Column
    {
        return new static(**$title**, $from);
    }
```

---
example to resolve: 
```php
public function columns(): array
{
    return [
        Column::make('Firstname', 'id')->sortable(), // <- change first make() call with title it doesn't display

/// now it would display the title Firstname on table.
        Column::make('Firstname', 'detail.firstname') // <- and I can use it further
    ];
}
```
And yet change in first make() call 'Firstname' -> back to 'ID' And it Worked. 
Firstname are displaying fine with all relationships which are needed.
```php
public function columns(): array
{
    return [
        Column::make('ID', 'id')->sortable(),
        Column::make('Firstname', 'detail.firstname')
    ];
}
```

I didn't understand but I found a workaround for me :)

Thanks for helping :)

**Reply by lrljoe** [reply]
If you're not using the "id" fields elsewhere, then use setAdditionalSelects in your configure method to get the relevant key fields.

Should then work smoothly

Your with() should come before query by the way, otherwise it goes a bit odd.

If not, if you localise the titles does it work?

**Reply by jurri** [reply]
I set it... 👍 
from my code: `->setAdditionalSelects(['users.id as id'])`

I cant use something before query(). 
If I do this I got this exeption: 'Call to undefined method Illuminate\Database\Eloquent\Builder::query()'

and for last ... yes.
If title was found for the first column return, it working for all others.

Thx
